tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -0,0 +1,470 @@
+    }
+  }
+
+  for (auto instr : fusion_instrs) {
+    if (!IsFusionSupported(*instr)) {
+      VLOG(2) << "Reject unsupported fusion instr " << instr->ToString();
+      continue;
+    } else if (!IsConsumerTheOnlyNonRootUser(*instr, *consumer)) {
+      VLOG(2) << "Reject maybe illegal instr " << instr->ToString()
+              << "; including it may create cycles in HLO.";
+      continue;
+    } else if (!IsProfitableFusionCandidate(*instr)) {
+      VLOG(2) << "Reject may-not-be profitable fusion instr "
+              << instr->ToString();
+      continue;
+    } else if (!HasOnlyRowMajorLayout(*instr)) {
+      VLOG(2) << "Reject non-row-major fusion instr " << instr->ToString();
+      continue;
+    } else {
+      fusion_instrs_.push_back(instr);
+    }
+  }
+
+  // Sort according to the number of outputs and instruciton counts, because
+  // we fuse only instructions with the same number of outputs and whose
+  // computations have the same instruction counts.
+  std::sort(fusion_instrs_.begin(), fusion_instrs_.end(),
+            [&](const HloInstruction* a, const HloInstruction* b) {
+              if (GetOutputSizeOfFusion(*a) == GetOutputSizeOfFusion(*b)) {
+                return a->fused_instruction_count() <
+                       b->fused_instruction_count();
+              }
+              return GetOutputSizeOfFusion(*a) < GetOutputSizeOfFusion(*b);
+            });
+}
+
+// Gets a next span of fusion instructions to be fused.
+absl::Span<HloInstruction*>
+HorizontalFusionImpl::FusionCandidates::GetNextSpanOfFusions() {
+  if (pos_ >= fusion_instrs_.size()) {
+    return absl::Span<HloInstruction*>();
+  }
+
+  // Fusing too many computations at a time may not be easily profitable and
+  // may increase compile time due to large kernels. Set a limit to it.
+  constexpr int64 kMaxFusionBatchSize = 32;
+  // CUDA has a parameter size limit of ~4k bytes.
+  constexpr int64 kMaxCudaParamSize = 4000;
+  size_t accum_io_size = 0;
+  auto reach_max_fusion_batch_size = [&](size_t left, size_t right) -> bool {
+    if (right - left >= kMaxFusionBatchSize) {
+      return true;
+    }
+
+    accum_io_size += fusion_instrs_.at(right)->fused_parameters().size() +
+                     GetOutputSizeOfFusion(*fusion_instrs_.at(right));
+
+    if (accum_io_size * 8 >= kMaxCudaParamSize) {
+      return true;
+    }
+
+    return false;
+  };
+
+  size_t left = pos_;
+  size_t right = pos_ + 1;
+  size_t first_output_size = GetOutputSizeOfFusion(*fusion_instrs_[left]);
+  for (; right < fusion_instrs_.size(); ++right) {
+    if (first_output_size != GetOutputSizeOfFusion(*fusion_instrs_[right])) {
+      // Cannot fuse computations who have different numbers of outputs.
+      break;
+    } else if (fusion_instrs_[left]->fused_instruction_count() !=
+               fusion_instrs_[right]->fused_instruction_count()) {
+      // Do not fuse computations of different instruction counts as it may
+      // introduce control divergence.
```

Comment:
The same number of instructions could still result in control divergence, right? It may be worth pointing out that this is a heuristic.

Reply:
Right. Revised the comments to make it clearer.